### ISSUE TITLE
Module concept and granularity / component approach

### ISSUE THREAD

**Issue by andreyvolokitin** [box/t3js#194]
[Docs state](http://t3js.org/docs/guides/thinking-in-t3#start-with-modules) that it is recommended for each module to cover as large an area of the page as possible. I'm trying to wrap my head around this concept while reflecting on my previous experience at the same time to figure out if I can fit t3js into my workflow. 

Usually, I handle most of the code with a component approach, which is quite granular. So i.e. the main navigation, a search box and drop-down menus inside a header are indeed separate components. Docs specifically state that these components should be a part of a larger "header" module, but there are times when such small granular components are reused in multiple places: header, footer, modal, other places. It makes sense to make them "independent"/separate, they usually represented by a DOM-element with some content, so they essentially are like modules in t3js ideology, except that they aren't.

I guess that the roots of the "make a module large" concept are coming from the need to delegate events and to "cover" as much interactive area as possible within module boundaries. I think I can understand why is that so (though currently, I'm not sure). But if we take this concept to the extreme, then we could just make one big "page" module to handle everything, and I'm not sure: on what DOM level this will become "bad practice" from "good practice". I.e. on the level of search box — it is a bad practice. On a level of the header — it is a good practice. But on a level of a page, this is again a bad practice. In reality, I think there will be some "intermediate" levels which may cause confusion as to if they should be a module boundary. And at the same time, the boundaries of a "component" are almost always clear, but there is no "component" concept in t3js and it is discouraged to make modules on a component level.

So in short: how to handle/reuse/manage small reusable components within t3js ideology?

**Comment by andreyvolokitin** [comment]
Another question: weren't there any considerations to change the way this works? So that it is not required to always use only bubblable events and big modules? Is it fundamentally possible in this ecosystem/framework/ideology to make such changes to the internals of T3? I'm asking so that maybe I could fork and adapt it to my needs, but that way I'm not sure if I can keep the benefits which framework provides

**Comment by zephraph** [comment]
You don't necessarily *have* to apply that principle rigorously. I think it's more a guideline than anything. Some of your examples, like the search bar, I treat as stand alone modules inside of a larger module (header). In general, there's nothing stopping you from having nested modules. The other pattern would be to have functionality (like search) as a *behavior* associated with the module. While this works, it implies that the behavior has to know something about the markup of the module it's associated with it... therefore I tend to avoid doing it that way.

**Comment by andreyvolokitin** [comment]
Yep, I realized that behaviors were probably created to address this same problem — to make it possible to reuse "components" within different modules. But instead of introducing "component" concept, we introducing the mechanism of sharing the module "delegation" logic across multiple modules, because for each component we should delegate its events to the containing module boundaries.

Currently, I'm trying to see if I indeed can use modules as granular components. One concern is performance because the page sometimes can contain quite a bit of something, and with small components, you are forced to avoid event delegation and to bind directly to each component. But simply because this guideline exists and the whole framework looks like built around this concept from the ground up, I think there might be some really strong reason for the "big components" guideline, but for the time being, I can't see it. 

BTW my thinking and approach to this is BEM-influenced, so I see "module" as a block. Ideally I want to make a lot of granular reusable modules/blocks with some elements inside them, and I am not sure how it will work in practice. I'm even not sure how it should work... I think I should experiment a bit and find out
